trunk/cpp/src/third-party/scipoptsuite-3.0.2/zimpl-3.3.1/doc/zimpl.man
.\" Process this file with
.\" groff -man -Tascii zimpl.man
.\"
.TH ZIMPL 1 "29 Jul 2012" Linux "User Manuals"
.SH NAME
zimpl \- Zuse Institute Mathematical Programming Language
.SH SYNOPSIS
.B zimpl  [ \-b ] [ \-D
.I name=value
.B ] [ \-f ] [ \-F
.I filter
.B ] [ \-h ] [ \-m ] [ \-l
.I length
.B [ \-n
.I cm|cn|cf
.B ] [ \-o
.I outfile
.B ] [ \-O ] [ \-P
.I filter
.B [ \-r ] [ \-t
.I lp|mps|hum|rlp|pip
.B ] [ \-s
.I seed
.B ] [ \-v
.I 0-5
.B ] [ \-V ]
.I file
.B ...
.SH DESCRIPTION
.B Zimpl
is a little language to translate the mathematical model of a problem
into a linear or (mixed-)integer mathematical program expressed in LPF
or MPS file format which can be read by a LP or MIP solver.

None of the options is necessary. All input files in 
.B zpl
format that are given are concatenated and processed as if they are
one big file. This makes is possible to combine different parts of a
model on the command line. In this case the first filename given will
be used as basename for the output files in absence of an 
.B \-o
option.

The 
.B \-t
option is to select the output format. 
.B mps 
can be read by every
solver. 
.B lpf 
is an CPLEX format, but can also read by several solvers.
BE aware that for example Range-rows can not be expressed in 
.B lpf . 
If you want to inspected the output yourself, 
.B hum
format is a variant of 
.B lpf
with no name mangeling whatever, but unfortunately this violates
several format restrictions, so no solver will read this in.
If
.B rlp
is specified the output is in 
.B lpf 
format, but rows and columns are randomly permuted.
 
.SH OPTIONS
.IP \-b
Enable 
.BR bison (1)
debugging output.
.IP "\-D name=value"
Sets the parameter
.I name
to the specified value. This is equivalent with having this line in the
Zimpl program: param name:=val .
.IP -f
Enable
.BR flex (1)
debugging output.
.IP "\-F filter"
The output is piped through a filter. A %s in the
string is replaced by the output filename. For example: 
-F "gzip \-c >%s.gz" would compress all the 
output files.
.IP -h
Show program help.
.IP "\-l length"
Sets the maximum length for variable names in LPF output files.
.IP -m
Write a CPLEX .mst Mip STart file.
.IP "\-n cm|cn|cf"
Select the format for the generation of constraint
names. Can be either 
.I cm 
which will number them 
1 ... n with a `c' in front. 
.I cn 
will use the name supplied in the subto statement and 
number them 1 ... n within the statement. 
.I cf 
will use the name given with the subto,
then a 1 .. n number like in 
.I cm 
and then append all the local variables from the forall statements.
.IP "-o outfile"
Sets the base-name for the output files. Default is the name of the
input file without extension.
.IP \-O 
Optimize the generated LP by doing some presolve analysis (experimental).
.IP "\-P filter" 
The input is piped through a filter. A %s in the
string is replaced by the input filename. For example: 
\-F "cpp -DWITH_C1 %s" would pass the input file through the
C-preprocessor.
.IP \-r 
Write a CPLEX .ord branching order file.
.IP "\-s seed"
Set the random number generator seed.
.IP "\-t lp|mps|hum|rlp|pip"
Selects the output format. Can be either 
.I lp
which is default, or 
.I mps 
, or
.I hum 
, which is (only) human readable, or 
.I rlp
, which is randomly permuted, or
.I pip
, which means polynimial IP.
.IP "\-v 0-5"
Set the verbosity level. 0 is quiet, 1 is default,
2 is verbose, 3 is chatter, and 5 is debug.
.IP \-V
Prints the version number.
.\".SH FILES
.\".SH ENVIRONMENT
.SH DIAGNOSTICS
The following diagnostics may be issued on stderr:

101 Bad filename
.RS
The name given with the 
.B \-o option 
is either missing, a directory name, or starts with a dot.
.RE
175 Illegal syntax for command line define -- ignored
.RS
The parameter of a 
.B \-D 
option was either not of the form 
.I "name=value"
or the name was not a legal identifier.
.RE
There are about a hundred other possible error messages, but these are
either about the contend of the input file, or system messages.
.SH BUGS
Please report any bug to the author.
.SH AUTHOR
Thorsten Koch <koch@zib.de>
.SH "WEBSITE"
The lastest version of Zimpl is always available at <http://zimpl.zib.de>.
.SH "SEE ALSO"
The Zimpl User Guide at <http://zimpl.zib.de/download/zimpl.pdf>.
.\".BR zpl (5),
.\".BR lpf (5),
.\".BR mps (5)





ChangeLog
2014-05-14  Ramon Casero  <rcasero@gmail.com>

	* matlab/CgalToolbox/CMakeLists.txt: (0.2.21)

	- Fix bug. cgal_meshseg and cgal_checl_self_intersect need to link
	against boost_thread.

	* CMakeLists.txt: (0.9.1)

	- Check that if we are in Windows, Gerardus is not on a network drive, as this
	will make CMake fail. As the error given by CMake will say nothing about the
	problem being the network drive, this is very confusing and annoying.

2014-05-13  Ramon Casero  <rcasero@gmail.com>

	* Add matlab/FileFormatToolbox/vmu2png.m: (0.1.0)

	- VMU2PNG Convert a microscope image file from .vmu to .png
	format.

2014-05-10  Ramon Casero  <rcasero@gmail.com>

	* matlab/FileFormatToolbox/scimat_save.m: (0.6.0)

	- Add output formats PNG (that allows saving images larger than
	4GB) with lossless compression, JPEG2000 and TIFF with
	AdobeDeflate compression.

2014-05-08  Ramon Casero  <rcasero@gmail.com>

	* matlab/FileFormatToolbox/scimat_load.m: (0.4.1)

	- Fix bug. Paul Cormack from Hamamatsu replied that pixel size is
	PhysicalWidth / PixelWidth, instead of PhysicalWidth /
	(PixelWidth-1).

	* matlab/FileFormatToolbox/scimat_load.m: (0.4.0)

	- Add new file format. Now we can read also .vmu (Hamamatsu
	Uncompressed Virtual Microscope Specimen) files.

	* matlab/FiltersToolbox/scinrrd_crop.m: (0.1.1)

	- When croppring, allow for a 4th dimension so that we can pass
	colour images with (row, col, slice, channel).

	- Update deprecated narginchk(), nargoutchk() syntax.

	* matlab/FileFormatToolbox/scinrrd_im2nrrd.m: (0.1.3)

	- Remove the limitation that input image must have 3 dimensions so
	that we can pass colour images with (row, col, slice, channel).

2014-05-07  Ramon Casero  <rcasero@gmail.com>

	* Rename matlab/FileFormatToolbox/scinrrd_load.m as matlab/FileFormatToolbox/scimat_load.m: (0.3.0)
	* Rename matlab/FileFormatToolbox/scinrrd_save.m as matlab/FileFormatToolbox/scimat_save.m: (0.5.0)
	* matlab/CardiacToolbox/process_misas_patient.m: (0.1.1)
	* matlab/FileFormatToolbox/scinrrd_im2nrrd.m: (0.1.2)
	* matlab/FileFormatToolbox/scinrrd_tiff2nrrd.m: (0.2.5)
	* matlab/FiltersToolbox/scinrrd_seg2voxel_stats.m: (0.1.2)
	* matlab/ItkToolbox/itk_imfilter.m: (0.7.7)
	* matlab/ItkToolbox/ItkImFilter.cpp: (1.6.2)
	* matlab/PointsToolbox/scimat_dmatrix_thickslice.m: (0.1.3)
	* matlab/test/test_itk_imfilter.m: (0.4.4)
	* matlab/test/test_cgal_closest_trifacet.m: (0.2.4)
	* matlab/test/test_pca_normalize.m: (0.1.1)
	* matlab/test/test_scimat_dmatrix_imgeodesic.m: (0.1.1)
	* matlab/test/test_surface_interp.m: (0.2.1)
	* matlab/test/test_cgal_trifacet_area.m: (0.1.1)
	* matlab/test/test_closest_trifacet.m: (0.3.2)

	- Renaming of scinrrd_load(), scinrrd_save() as scimat_load(),
	scimat_save(), respectively.

	- While we are at it, update deprecated narginchk(), nargoutchk()
	syntax when needed, and change function name to lower case in the
	help header.

	* matlab/test/test_surface_param.m: (0.6.1)

	- Renaming of scinrrd_load(), scinrrd_save() as scimat_load(),
	scimat_save(), respectively.

	- Check for self-intersections.


2014-05-06  Ramon Casero  <rcasero@gmail.com>

	* matlab/ManifoldToolbox/tri_sphparam.m: (0.3.0)
	* matlab/test/test_tri_sphparam.m: (0.1.1)

	- Fix bug. We had broken the MDS method, by always producing an
	array of NaNs as the output parametrization. We only need to check
	whether we want to replace the found solution by the initial guess
	in the SMACOF-based methods (because in MDS, there's no initial
	guess).

	- Enforce that for the MDS method, there can be no initial guess.

	- Add new output variable that tells whether the output
	parametrization is valid (all tetrahedra with positive area).

	- In the MDS method, avoid recomputing the sphere radius. We stick
	to what the user provided, or the estimate based on the mesh area.

	- All methods use a matrix of chord distances instead of geodesic
	distances, so move the code to do the conversion out of each
	"case" in the "switch" statement, to have only one call for all
	methods.

	- Input variable Y is now Y0, which makes more sense and
	simplifies the code a bit.

2014-05-05  Ramon Casero  <rcasero@gmail.com>

	* matlab/ManifoldToolbox/tri_sphparam.m: (0.2.0)

	- Fix the way we select the solution we return. Now we checked
	whether the initial guess and found best solution are valid. If
	both are, then we return the one with less stress. If only one is
	valid, we return that one. If neither the initial guess is valid
	nor a valid solution could be found, we return an array of NaNs.

	- Change output syntax. Now return stress value of the initial
	guess as a separate variable.

2014-05-01  Ramon Casero  <rcasero@gmail.com>

	* matlab/test/test_cons_smacof_pip.m: (0.2.5)

2014-04-29  Ramon Casero  <rcasero@gmail.com>

	* matlab/PointsToolbox/cons_smacof_pip.m: (0.3.5)

	- Fix bug. There's a bug in scip-3.1.0.linux.x86_64.gnu.opt.spx,
	which causes the solution to be written as an empty file when the
	quiet flag is used. As a workaround, we disable the quiet flag in
	that case, and instead send the output to /dev/null, as suggested
	by Stefan Vigerske.

2014-04-22  Ramon Casero  <rcasero@gmail.com>

	* Add matlab/FileFormatToolbox/ndpi_focus_points.m: (0.1.0)

	- NDPI_FOCUS_POINTS Read focus points coordinates from .ndpi image
	file obtained from a Hamamatsu microscope.

2014-04-07  Ramon Casero  <rcasero@gmail.com>

	* Add matlab/FiltersToolbox/amrf_seg.m: (0.1.0)

	- AMRF_SEG  Adaptive Markov Random Field segmentation.

	* Add matlab/FiltersToolbox/im_modes.m: (0.1.0)

	- IM_MODES Estimate typical values (modes) of
	foreground/background intensities in a greyscale image.

2014-03-16  Ramon Casero  <rcasero@gmail.com>

	* matlab/ManifoldToolbox/tri_sphparam.m: (0.1.1)

	- Classical MDS solution is invariant under a mirroring
	transformation. That is, if we had the ground truth, a mirror
	image over the longitude coordinate produces the same stress, but
	with all the triangle orientations inverted. Now we mirror the

tags/release-0.10.0/trunk/cpp/src/third-party/scipoptsuite-3.0.2/zimpl-3.3.1/src/zimpl.c
/* $Id: zimpl.c,v 1.93 2012/07/29 15:09:31 bzfkocht Exp $ */
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
/*                                                                           */
/*   File....: zimpl.c                                                       */
/*   Name....: Zuse Institute Mathematical Programming Language              */
/*   Author..: Thorsten Koch                                                 */
/*   Copyright by Author, All rights reserved                                */
/*                                                                           */
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
/*
 * Copyright (C) 2001-2012 by Thorsten Koch <koch@zib.de>
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
 */

#include <sys/types.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <ctype.h>
#include <string.h>
#include <errno.h>

#include "lint.h"
#include "bool.h"
#include "mshell.h"
#include "stkchk.h"
#include "random.h"
#include "blkmem.h"
#include "ratlptypes.h"
#include "numb.h"
#include "elem.h"
#include "tuple.h"
#include "mme.h"
#include "set.h"
#include "symbol.h"
#include "define.h"
#include "bound.h"
#include "mono.h"
#include "term.h"
#include "stmt.h"
#include "local.h"
#include "list.h"
#include "entry.h"
#include "conname.h"
#include "xlpglue.h"
#include "zlpglue.h"
#include "prog.h"
#include "metaio.h"
#include "strstore.h"
#include "zimpllib.h"

extern int yydebug;
extern int yy_flex_debug;

static const char* const options = "bD:fF:hl:mn:o:OP:rs:t:v:V";
static const char* const usage   = "usage: %s [options] file ...\n";
static const char* const title   = "This file was automatically generated by Zimpl";

static const char* const help =
"\n" \
"  -b             enable bison debugging output.\n" \
"  -D name=value  assign value to parameter name.\n" \
"  -f             enable flex debugging output.\n" \
"  -F filter      filter output, for example \"gzip -c >%%s.gz\"\n" \
"  -h             show this help.\n" \
"  -l length      Maximum length of names in output file.\n" \
"  -m             write CPLEX MIP start value file.\n"
"  -n cm|cn|cf    name constraint make/name/full\n" \
"  -o outfile     select name for the output file. Default is the name of\n" \
"                 the input file without extension.\n" \
"  -P cmd         Pipe input through command, e.g. \"cpp -DONLY_X %%s\"\n" \
"  -r             write CPLEX branching order file.\n" \
"  -s seed        random number generator seed.\n" \
"  -t lp|mps|hum|rlp|pip  select output format. Either LP (default), MPS format,\n" \
"                 human readable HUM, randomly permuted LP, or PIP polynomial IP.\n" \
"  -v[0-5]        verbosity level: 0 = quiet, 1 = default, up to 5 = debug\n" \
"  -V             print program version\n" \
"  filename       is the name of the input ZPL file.\n" \
"\n" ; 

/* "  -O             optimize LP by preprocessing.\n" \
 */
#ifdef WITH_CALLTRACE /* Does only work with gcc */
void __cyg_profile_func_enter(void *this_fn, void *call_site) __attribute__((no_instrument_function));
void __cyg_profile_func_exit(void *this_fn, void *call_site) __attribute__((no_instrument_function));
static int call_depth = 0;

void __cyg_profile_func_enter(void *this_fn, void *call_site)
{
   call_depth++;
   
   fprintf(stderr, "%p %d\n", this_fn, call_depth);
}

void __cyg_profile_func_exit(void *this_fn, void *call_site) 
{
   call_depth--;
}
#endif

static char* add_extention(const char* filename, const char* extension)
{
   char* basename;
   
   assert(filename != NULL);
   assert(strlen(filename) > 0);
   assert(extension != NULL);

   basename = malloc(strlen(filename) + strlen(extension) + 1);

   assert(basename != NULL);

   strcpy(basename, filename);

   /*lint -e{429} suppress 'Custodial pointer 'basename' (line 96) has not been freed or returned' */
   return strcat(basename, extension);
}

static const char* strip_path(const char* filename)
{
   const char* s;
   
   assert(filename != NULL);
   assert(strlen(filename) > 0);
   
   s = strrchr(filename, DIRSEP);

   return (s == NULL) ? filename : s + 1;
}

static char* strip_extension(char* filename)
{
   int i;
   
   assert(filename != NULL);
   assert(strlen(filename) > 0);

   for(i = (int)strlen(filename) - 1; i >= 0; i--)
      if (filename[i] == DIRSEP || filename[i] == '.')
         break;

   if (i >= 0 && filename[i] == '.')
      filename[i] = '\0';

   i = (int)strlen(filename);
   
   if (i == 0 || filename[i - 1] == DIRSEP)
   {
      fprintf(stderr, "*** Error 101: Bad filename\n");
      exit(EXIT_FAILURE);
   }
   return filename;
}

static void check_write_ok(FILE* fp, const char* filename)
{
   if (ferror(fp))
   {
      fprintf(stderr, "*** Error 102: File write error\n");
      perror(filename);
   }
}

int main(int argc, char* const* argv)
{
   Prog*         prog;
   Set*          set;
   void*         lp;
   const char*   extension = "";
   char*         filter    = strdup("%s");
   char*         outfile;
   char*         tblfile;
   char*         ordfile;
   char*         mstfile;
   char*         basefile = NULL;
   char*         inppipe  = NULL;
   char*         outpipe;
   LpFormat      format   = LP_FORM_LPF;
   FILE*         fp;
   Bool          write_order = FALSE;
   Bool          write_mst   = FALSE;
   Bool          presolve    = FALSE;
   int           name_length = 0;
   char*         prog_text;
   unsigned long seed = 13021967UL;
   char**        param_table;
   int           param_count = 0;
   int           c;
   int           i;
   FILE*         (*openfile)(const char*, const char*) = fopen;
   int           (*closefile)(FILE*)                   = fclose;

   stkchk_init();
   
   yydebug       = 0;
   yy_flex_debug = 0;
   verbose       = VERB_NORMAL;
   param_table   = malloc(sizeof(*param_table));
   
   while((c = getopt(argc, argv, options)) != -1)
   {
      switch(c)
      {
      case 'b' :
         yydebug = 1;
         break;
      case 'D' :
         param_table =
            realloc(param_table, ((unsigned int)param_count + 1) * sizeof(*param_table));
         param_table[param_count] = strdup(optarg);
         param_count++;
         break;
      case 'h' :
         zpl_print_banner(stdout, TRUE);
         printf(usage, argv[0]);
         puts(help);
         exit(0);
      case 'f' :
         yy_flex_debug = 1;
         break;
      case 'F' :
         free(filter);
         
         filter    = strdup(optarg);
         openfile  = popen;
         closefile = pclose;
         break;
      case 'l' :
         name_length = atoi(optarg);
         break;
      case 'm' :
         write_mst = TRUE;
         break;
      case 'n' :
         if (*optarg != 'c')
         {
            fprintf(stderr, usage, argv[0]);
            exit(0);
         }
         switch(optarg[1])
         {
         case 'm' :
            conname_format(CON_FORM_MAKE);
            break;
         case 'n' :
            conname_format(CON_FORM_NAME);
            break;
         case 'f' :
            conname_format(CON_FORM_FULL);
            break;
         default :
            fprintf(stderr, usage, argv[0]);
            exit(0);
         }
         break;
      case 'o' :
         basefile = strdup(optarg);
         break;
      case 'O' :
         presolve = TRUE;
         break;
      case 'P' :
         inppipe = strdup(optarg);
         break;
      case 's' :
         seed = (unsigned long)atol(optarg);
         break;
      case 'r' :
         write_order = TRUE;
         break;
      case 't' :
         switch(tolower(*optarg))
         {
         case 'h' :
            format = LP_FORM_HUM;
            break;
         case 'm' :
            format = LP_FORM_MPS;
            break;
         case 'l' :
            format = LP_FORM_LPF;
            break;
         case 'p' :
            format = LP_FORM_PIP;
            break;
         case 'r' :
            format = LP_FORM_RLP;
            break;
         default :
            if (verbose > VERB_QUIET)
               fprintf(stderr,
                  "--- Warning 103: Output format \"%s\" not supported, using LP format\n",
                  optarg);
            format = LP_FORM_LPF;
            break;
         }
         break;
      case 'v' :
         verbose = atoi(optarg);
         break;
      case 'V' :
         printf("%s\n", VERSION);
         exit(0);
      case '?':
         fprintf(stderr, usage, argv[0]);
         exit(0);
      default :
         abort();
      }
   }
   if ((argc - optind) < 1)
   {
      fprintf(stderr, usage, argv[0]);      
      exit(0);
   }

   zpl_print_banner(stdout, TRUE);
   
   if (basefile == NULL)
      basefile = strip_extension(strdup(strip_path(argv[optind])));

   switch(format)
   {
   case LP_FORM_LPF :
      extension = ".lp";
      break;
   case LP_FORM_MPS :
      extension = ".mps";
      break;
   case LP_FORM_HUM :
      extension = ".hum";
      break;
   case LP_FORM_RLP :
      extension = ".rlp";
      break;
   case LP_FORM_PIP :
      extension = ".pip";
      break;
   default :
      abort();
   }
   assert(extension != NULL);

   outfile = add_extention(basefile, extension);
   tblfile = add_extention(basefile, ".tbl");
   ordfile = add_extention(basefile, ".ord");
   mstfile = add_extention(basefile, ".mst");
   
   outpipe = malloc(strlen(basefile) + strlen(filter) + 256);

   assert(outpipe != NULL);

   blk_init();
   str_init();
   rand_init(seed);
   numb_init(TRUE);
   elem_init();
   set_init();
   mio_init();
   interns_init();
   local_init();
   
   /* Make symbol to hold entries of internal variables
    */
   set = set_pseudo_new();
   (void)symbol_new(SYMBOL_NAME_INTERNAL, SYM_VAR, set, 100, NULL);
   set_free(set);
   
   /* Now store the param defines
    */
   for(i = 0; i < param_count; i++)
      zpl_add_parameter(param_table[i]);

   /* Next we read in the zpl program(s)
    */
   prog = prog_new();

   for(i = optind; i < argc; i++)
      prog_load(prog, inppipe, argv[i]);

   if (prog_is_empty(prog))
   {
      fprintf(stderr, "*** Error 168: No program statements to execute\n");
      exit(EXIT_FAILURE);
   }
   if (verbose >= VERB_DEBUG)
      prog_print(stderr, prog);
   
   lp = xlp_alloc(argv[optind], write_mst || write_order, NULL);
   zlp_setnamelen(lp, name_length);
   
   prog_execute(prog, lp);

   /* Presolve
    */
   if (presolve)
      fprintf(stderr, "--- Warning: Presolve no longer support. If you need it, send me an email\n");
#if 0 
      if (!zlp_presolve())
         exit(EXIT_SUCCESS);
#endif
   if (verbose >= VERB_NORMAL)
      zlp_stat(lp);
   
   /* Write order file 
    */
   if (write_order)
   {
      sprintf(outpipe, filter, ordfile, "ord");

      if (verbose >= VERB_NORMAL)
         printf("Writing [%s]\n", outpipe);

      if (NULL == (fp = (*openfile)(outpipe, "w")))
      {
         fprintf(stderr, "*** Error 104: File open failed ");
         perror(ordfile);
         exit(EXIT_FAILURE);
      }
      zlp_orderfile(lp, fp, format);
         
      check_write_ok(fp, ordfile);
         
      (void)(*closefile)(fp);
   }
   /* Write MST file 
    */
   if (write_mst)
   {
      sprintf(outpipe, filter, mstfile, "mst");

      if (verbose >= VERB_NORMAL)
         printf("Writing [%s]\n", outpipe);

      if (NULL == (fp = (*openfile)(outpipe, "w")))
      {
         fprintf(stderr, "*** Error 104: File open failed ");
         perror(mstfile);
         exit(EXIT_FAILURE);
      }
      zlp_mstfile(lp, fp, format);
         
      check_write_ok(fp, mstfile);
         
      (void)(*closefile)(fp);
   }
   /* Write Output
    */
   sprintf(outpipe, filter, outfile, "lp");

   if (verbose >= VERB_NORMAL)
      printf("Writing [%s]\n", outpipe);

   if (NULL == (fp = (*openfile)(outpipe, "w")))
   {
      fprintf(stderr, "*** Error 104: File open failed ");
      perror(outfile);
      exit(EXIT_FAILURE);
   }
   if (format != LP_FORM_RLP)
      prog_text = prog_tostr(prog, format == LP_FORM_MPS ? "* " : "\\ ", title, 128);
   else
   {
      prog_text = malloc(strlen(title) + 4);
      
      assert(prog_text != NULL);

      sprintf(prog_text, "\\%s\n", title);
   }
   zlp_write(lp, fp, format, prog_text);

   check_write_ok(fp, outfile);

   (void)(*closefile)(fp);

   /* We do not need the translation table for human readable format
    * Has to be written after the LP file, so the scaling has been done.
    */
   if (format != LP_FORM_HUM)
   {
      /* Write translation table
       */
      sprintf(outpipe, filter, tblfile, "tbl");

      if (verbose >= VERB_NORMAL)
         printf("Writing [%s]\n", outpipe);

      if (NULL == (fp = (*openfile)(outpipe, "w")))
      {
         fprintf(stderr, "*** Error 104: File open failed");
         perror(tblfile);
         exit(EXIT_FAILURE);
      }
      zlp_transtable(lp, fp, format);

      check_write_ok(fp, tblfile);

      (void)(*closefile)(fp);
   }




   free(prog_text);
   
   if (verbose >= VERB_DEBUG) 
      symbol_print_all(stderr);

#if defined(__INSURE__) || !defined(NDEBUG) || defined(FREEMEM)
   
   /* Now clean up. 
    */
   if (inppipe != NULL)
      free(inppipe);
   
   for(i = 0; i < param_count; i++)
      free(param_table[i]);
   free(param_table);
   
   prog_free(prog);

   local_exit();
   interns_exit();
   xlp_free(lp);
   mio_exit();
   symbol_exit();
   define_exit();
   set_exit();
   elem_exit();
   numb_exit();
   str_exit();
   blk_exit();
   
   free(mstfile);
   free(ordfile);
   free(outfile);
   free(tblfile);
   free(basefile);
   free(filter);
   free(outpipe);

   if (verbose >= VERB_NORMAL)
   {
      mem_display(stdout);
      stkchk_maximum(stdout);
   }
#endif /* __INSURE__ || !NDEBUG || FREEMEM */
   return 0;
}







